File: examples/server-dart/counter.dart
import 'package:orca_engine/orca_engine.dart';

// ── Home Page ─────────────────────────────────────────────

final homePage = PageDefinition.create(PageDefinitionConfig(
  id: 'home',
  title: 'Counter',
  state: (_) => [
    const StateDefinition(key: 'count', scope: 'page', initial: 0),
  ],
  render: (ctx, info) => Scaffold(
    appBar: AppBar(
      title: Text(data: 'Counter Example'),
      centerTitle: true,
    ),
    body: Center(
      child: Column(
        mainAxisAlignment: 'center',
        children: [
          Text(
            data: V.transform(V.pageState('count'), [TV.toString$()]),
            style: {'fontSize': 48, 'fontWeight': 'bold'},
          ),
          SizedBox(height: 24),
          Row(
            mainAxisAlignment: 'center',
            children: [
              ElevatedButton(
                child: Text(data: '-'),
                actions: {
                  'onTap': setState('count',
                      V.transform(V.pageState('count'), [TV.subtract(V.static$(1))])),
                },
              ),
              SizedBox(width: 16),
              ElevatedButton(
                child: Text(data: '+'),
                actions: {
                  'onTap': setState('count',
                      V.transform(V.pageState('count'), [TV.add(V.static$(1))])),
                },
              ),
            ],
          ),
          SizedBox(height: 16),
          ElevatedButton(
            child: Text(data: 'Reset'),
            actions: {
              'onTap': sequential([
                setState('count', V.static$(0)),
                showSnackbar('Counter reset to 0'),
              ]),
            },
          ),
        ],
      ),
    ),
  ),
));

// ── App ──────────────────────────────────────────────────

final homeFlow = Flow.create(FlowConfig(
  name: 'home',
  routes: [RouteDefinition(path: 'home', page: homePage)],
));

final counterApp = App.create(AppConfig(
  id: 'counter',
  name: 'Counter',
  flows: [homeFlow],
));

File: examples/server/counter.ts
import { App, Flow, PageDefinition } from "../../engine/src/core";
import { V, SetState, ShowSnackbar, Sequential } from "../../engine/src/types";
import {
  Scaffold,
  AppBar,
  Column,
  Row,
  Text,
  ElevatedButton,
  Center,
  SizedBox,
} from "../../engine/src/components";

// ── Home Page ─────────────────────────────────────────────

const homePage = PageDefinition.create({
  id: "home",
  title: "Counter",
  state: [{ key: "count", scope: "page", initial: 0 }],
  render: (_ctx, _info) =>
    Scaffold.new({
      appBar: AppBar.new({
        title: Text.new({ data: "Counter Example" }),
        centerTitle: true,
      }),
      body: Center.new({
        child: Column.new({
          mainAxisAlignment: "center",
          children: [
            Text.new({
              data: V.transform(V.pageState("count"), [{ type: "toString" }]),
              style: { fontSize: 48, fontWeight: "bold" },
            }),
            SizedBox.new({ height: 24 }),
            Row.new({
              mainAxisAlignment: "center",
              children: [
                ElevatedButton.new({
                  child: Text.new({ data: "-" }),
                  actions: {
                    onTap: SetState(
                      "count",
                      V.transform(V.pageState("count"), [
                        { type: "subtract", by: V.static(1) },
                      ]),
                    ),
                  },
                }),
                SizedBox.new({ width: 16 }),
                ElevatedButton.new({
                  child: Text.new({ data: "+" }),
                  actions: {
                    onTap: SetState(
                      "count",
                      V.transform(V.pageState("count"), [
                        { type: "add", by: V.static(1) },
                      ]),
                    ),
                  },
                }),
              ],
            }),
            SizedBox.new({ height: 16 }),
            ElevatedButton.new({
              child: Text.new({ data: "Reset" }),
              actions: {
                onTap: Sequential(
                  SetState("count", V.static(0)),
                  ShowSnackbar("Counter reset to 0"),
                ),
              },
            }),
          ],
        }),
      }),
    }),
});

// ── App ──────────────────────────────────────────────────

const homeFlow = Flow.create({
  name: "home",
  routes: [{ path: "home", page: homePage }],
});

export const counterApp = App.create({
  id: "counter",
  name: "Counter",
  flows: [homeFlow],
});
